### PR TITLE
Add mission set current item feature

### PR DESCRIPTION
**Pull request by Katawann**
this feature has been removed in commit c6f67067c160759b4a738aabf3a9443e0e48c055. It allows to set the current item in a mission (async)

related to issue #1003

### CODE REVIEW

File: src/core/mavlink_mission_transfer.cpp [hunk 1]
```diff
@@ -767,4 +775,126 @@ void MAVLinkMissionTransfer::ClearWorkItem::callback_and_reset(Result result)
...
+}
+
+void MAVLinkMissionTransfer::SetCurrentWorkItem::start()
+{
+    std::lock_guard<std::mutex> lock(_mutex);
+
+    _started = true;
+
+    // If we coudln't find it, the requested item is out of range and probably an invalid argument.
+    if (_current < 0) {
```
Comment (julianoes) [inline src/core/mavlink_mission_transfer.cpp]: We should also check for greater than length of items. Although never mind, we don't have that information in this class, the autopilot will have to take care of it.

File: src/core/mavlink_mission_transfer.cpp [hunk 2]
```diff
@@ -767,4 +775,126 @@ void MAVLinkMissionTransfer::ClearWorkItem::callback_and_reset(Result result)
...
+        callback_and_reset(Result::CurrentInvalid);
```
Comment (julianoes) [inline src/core/mavlink_mission_transfer.cpp]: This error was initially for something else but I guess it fits.

File: src/core/mavlink_mission_transfer.cpp [hunk 3]
```diff
@@ -767,4 +775,126 @@ void MAVLinkMissionTransfer::ClearWorkItem::callback_and_reset(Result result)
...
+
+void MAVLinkMissionTransfer::SetCurrentWorkItem::process_mission_current(const mavlink_message_t& message)
+{
+    std::lock_guard<std::mutex> lock(_mutex);
+
+    mavlink_mission_current_t mission_current;
+    mavlink_msg_mission_current_decode(&message,&mission_current);
+
+    _timeout_handler.remove(_cookie);
+    _current = mission_current.seq;
```
Comment (julianoes) [inline src/core/mavlink_mission_transfer.cpp]: I think we should instead check if the returned current is correct and report success if so.

File: src/core/mavlink_mission_transfer.cpp [hunk 4]
```diff
@@ -767,4 +775,126 @@ void MAVLinkMissionTransfer::ClearWorkItem::callback_and_reset(Result result)
...
+    std::lock_guard<std::mutex> lock(_mutex);
+
+    mavlink_statustext_t status;
+    mavlink_msg_statustext_decode(&message,&status);
+
+    _timeout_handler.remove(_cookie);
+    //TODO status.severity
+
+    callback_and_reset(Result::ProtocolError);
+}
```
Comment (julianoes) [inline src/core/mavlink_mission_transfer.cpp]: Why is this required? I don't think a status text output should concern us.

File: src/core/mavlink_mission_transfer.h [hunk 5]
```diff
@@ -229,6 +262,8 @@ class MAVLinkMissionTransfer {
 
     void clear_items_async(uint8_t type, ResultCallback callback);
 
+    void set_current_item_async(uint8_t type, int &current, ResultCallback callback);
```
Comment (julianoes) [inline src/core/mavlink_mission_transfer.h]: I wouldn't use a reference here. Just pass it by value.

File: src/core/mavlink_mission_transfer.cpp (hunk 4 above)
Comment (Katawann) [reply src/core/mavlink_mission_transfer.cpp]: From the docs they say:

> Drone (server) receives message and attempts to update the current mission sequence number.
> 
> - On success, the Drone should broadcast a MISSION_CURRENT message containing the current sequence number (seq).
> - On failure, the Drone should broadcast a STATUSTEXT with a MAV_SEVERITY and a string stating the problem. This may be displayed in the UI of receiving systems.

So I was not sure if that something we need to check and how to do it

File: src/core/mavlink_mission_transfer.cpp (hunk 1 above)
Comment (Katawann) [reply src/core/mavlink_mission_transfer.cpp]: It was my thought but I did not find a way to check it unless we receive the item in parameters

**Comment by Katawann** [comment]
> so either you or I should add unit tests 
Yep agree! I will not have the time in next days but it can be good practice for me to do it. I will commit here when it's done !

**Comment by julianoes** [comment]
Oh, and we should also use this function in one of the integration tests, that way we're less likely to break it again.

**Comment by Katawann** [comment]
Oups not a review I wanted to do...

Not working but I think I will need your help to perform the unit test. I'm not used to this function and I'm running out of time to do it properly. For now it fails because it blocks on the EXPECT_EQ. What I missing ?

**Comment by julianoes** [comment]
@Katawann I've added some commits. Let's see if this passes now.

**Comment by Katawann** [comment]
Great ! Thank you for your help. I will check what you've done and do it with the upload mission raw I will pull request in few times !

**Comment by julianoes** [comment]
@Katawann the upload `mission_raw` feature should be straightforward, basically a copy from `mission` and doesn't have any tests. Although an integration test would be nice :).

**Comment by julianoes** [comment]
I realized that two more unit tests for retransmission and timeout need to be added. I'm on that.

**Comment by julianoes** [comment]
Unrelated CI error, merging.
